vignettes/loo2-moment-matching.Rmd

```diff
@@ -0,0 +1,301 @@
+
+```{r, stancode}
+stancode = "
+data {
+  int<lower=1> K;
+  int<lower=1> N;
+  matrix[N,K] x;
+  int y[N];
+  vector[N] offset;
+
+  real beta_prior_scale;
+  real alpha_prior_scale;
+}
+parameters {
+  vector[K] beta;
+  real intercept;
+}
+model {
+  y ~ poisson(exp(x * beta + intercept + offset));
+  beta ~ normal(0,beta_prior_scale);
+  intercept ~ normal(0,alpha_prior_scale);
+}
+generated quantities {
+  vector[N] log_lik;
+  for (n in 1:N)
+    log_lik[n] = poisson_lpmf(y[n] | exp(x[n] * beta + intercept + offset[n]));
+}
+"
+```
+
+Following the usual approach recommended in 
+[_Writing Stan programs for use with the loo package_](http://mc-stan.org/loo/articles/loo2-with-rstan.html), 
+we compute the log-likelihood for each observation in the 
+`generated quantities` block of the Stan program.
+
+
+## Setup
+
+In addition to __loo__, we load the __rstan__ package for fitting the model,
+and the __rstanarm__ package for the data.
+
+```{r, setup, message=FALSE}
+library("rstan")
+library("loo")
+library("rstanarm")
+seed <- 9547
+set.seed(seed)
+```
+
+## Fitting the model with RStan
+
+Next we fit the model in Stan using the __rstan__ package:
+
+```{r, modelfit}
+# Prepare data
+data(roaches)
+roaches$roach1 <- sqrt(roaches$roach1)
+y <- roaches$y
+x <- roaches[,2:4]
+offset <- log(roaches[,5])
+n <- dim(x)[1]
+k <- dim(x)[2]
+
+standata <- list(N = n, K = k, x = as.matrix(x), y = y,offset=offset,beta_prior_scale = 2.5,alpha_prior_scale = 5.0)
+
+# Compile
+stanmodel <- stan_model(model_code = stancode)
+
+# Fit model
+fit <- sampling(stanmodel, data = standata, seed = seed, refresh = 0)
+print(fit, pars = "beta")
+```
+
+Let us now evaluate the predictive performance of the model using `loo()`.
+
+```{r, loo1}
+loo1 <- loo(fit)
+loo1
+```
+
+We notice that the `loo()` output warns us that there are some observations
+which are very influential, and thus the
+accuracy of importance sampling is compromised as indicated by the large Pareto $k$
+diagnostic values (> 0.7). As discussed in the demo
+[_Bayesian data analysis - roaches cross-validation demo_](https://avehtari.github.io/modelselection/roaches.html),
+this may be an indication of model misspecification. Despite that, it is still beneficial
+to be able to evaluate the predictive performance of the model accurately.
+
+## Moment matching correction for importance sampling
+
+To improve the accuracy of the `loo()` result above, we could perform
+leave-one-out cross-validation by explicitly leaving out single observations
+and refitting the model repeatedly. However, the Pareto $k$ diagnostics
+indicate that there are 19 observations which are problematic. This would
+require 19 model refits which requires lots of computation time.
+
+Instead of refitting, we can perform a moment matching correction to
+the importance sampling for the problematic observations, avoiding refitting
+the model. This can be done with the `loo_moment_match()` function from the
```

Comment:
remove ", avoiding refitting the model" (it's already said in the beginning of the sentence)

function from -> function in